docs/_docs/20-fasterkv-basics.md

```diff
@@ -107,7 +107,95 @@ Apart from Key and Value, the IFunctions interface is defined on three additiona
+  - *InitialUpdater*
+    - `NeedInitialUpdate`
+      - ***Required Length***: FASTER sets updateInfo.FullValueLength and updateInfo.UsedValueLength on input to the VarLengthStructSettings result or revivification info. The app sets it to the required length if necessary. FASTER will allocate a record with the required length and call `InitialUpdater`.
+      - ***Expiration***: The app sets updateInfo.DeleteRecord true and returns false. FASTER will perform a logical delete of the record by continuing with InitialUpdater, and RMW will return (!status.Found && status.CreatedRecord).
+      - ***Cancellation***: The app sets updateInfo.CancelOperation True and returns false. RMW() returns (!status.Found).
+    - `InitialUpdater`
+      - ***Required Length***: FASTER sets updateInfo.FullValueLength and updateInfo.UsedValueLength on input to the values previously obtained.
+        - If the length is sufficient, the app sets updateInfo.UsedValueLength to the used length and returns true.
+        - Otherwise, the app sets updateInfo.UsedValueLength to the required length and returns false. RMW will return RETRY_NOW and FASTER will retry the operation.
+      - ***Expiration***: The app sets updateInfo.DeleteRecord true and returns false. FASTER will perform a logical delete of the record by continuing with the CAS, and RMW will return (!status.Found && status.CreatedRecord).
+      - ***Cancellation***: The app sets updateInfo.CancelOperation True and returns false. RMW() returns (!status.Found).
+    - `InPlaceUpdater`
+      - ***Required Length***: FASTER sets updateInfo.FullValueLength and updateInfo.UsedValueLength on input to the values previously obtained.
+        - If the length is sufficient, the app sets updateInfo.UsedValueLength to the used length and returns true.
+        - Otherwise, the app sets updateInfo.UsedValueLength to the required length and returns false. FASTER will allocate a record with the required length and call `InitialUpdater`.
+      - ***Expiration***: The app sets updateInfo.DeleteRecord true and returns false. FASTER will perform a logical delete of the record by in-place update, and RMW will return (!status.Found).
+      - ***Cancellation***: The app sets updateInfo.CancelOperation True and returns false. RMW() returns (!status.Found).
+  - *CopyUpdater*
+    - `NeedCopyUpdate`
+      - ***Required Length***: FASTER sets updateInfo.FullValueLength and updateInfo.UsedValueLength on input to the VarLengthStructSettings result or revivificatgion info. The app sets it to the required length if necessary. FASTER will allocate a record with the required length and call `CopyUpdater`.
+      - ***Expiration***: The app sets updateInfo.DeleteRecord true and returns false. FASTER will perform a logical delete of the record by calling InitialUpdater, and RMW will return (!status.Found && status.CreatedRecord).
```

Comment:
can we allow the user to expire the data AND cancel the operation. so we are deleting the record by adding a tombstone, and we will not call IU.

Reply:
This comment applies to InPlaceUpdater, not CopyUpdater, which does not go to the IU path. 

Doc updated; if IPU sets both RMWInfo.CancelOperation and RMWInfo.DeleteRecord, then RMW sets the tombstone and returns (status.NotFound | status.Record.Created). Otherwise, .DeleteRecord moves to InitialUpdater.

The app may do its own internal "expire and reset" logic within the `InPlaceUpdater` call and return true, which to FASTER is indistinguishable from any other successful `InPlaceUpdater` call.